lens/lens.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"github.com/sourcenetwork/immutable/enumerable"
+
+	"github.com/sourcenetwork/defradb/client"
+)
+
+type schemaVersionID = string
+
+type LensDoc = map[string]any
+
+type lensInput struct {
+	SchemaVersionID schemaVersionID
+	Doc             LensDoc
+}
+
+type Lens interface {
+	enumerable.Enumerable[LensDoc]
+	Put(schemaVersionID schemaVersionID, value LensDoc) error
+}
+
+type lens struct {
+	lensRegistry client.LensRegistry
+
+	lensPipesBySchemaVersionIDs      map[schemaVersionID]enumerable.Concatenation[LensDoc]
+	lensInputPipesBySchemaVersionIDs map[schemaVersionID]enumerable.Queue[LensDoc]
+	outputPipe                       enumerable.Concatenation[LensDoc]
+
+	schemaVersionHistory map[schemaVersionID]*targetedSchemaHistoryLink
+
+	source enumerable.Queue[lensInput]
+}
+
+var _ Lens = (*lens)(nil)
+
+func New(
```

Comment:
todo: Public methods should be documented.

Reply:
Will make private

- [ ] Make private